### ISSUE TITLE
CLion Configuration

### ISSUE THREAD

**Issue by surahman** [ClickHouse/ClickHouse#34736]
I have seen a number of questions on this topic both on GH and one on SO without any resolution, so I figured I would ask whilst providing more details. I am trying to configure CLion for builds and debugging with the Clickhouse source. I am currently debugging with `LLDB` on the CLI and it is painfully slow.

I am trying to build and attach the debugger to the `clickhouse-server` target (`--target clickhouse-server`). I am able to build the executable but the IDE warns me that it cannot hit the breakpoints. I believe the debugger is unable to attach to the executable despite it having the debugging symbols loaded. The Clickhouse client also warns me that the server is running in debug mode, does this confirm it was built with debug symbols set or is this just more verbose logging? I also have to manually set the Executable in Run/Debug Configurations for `clickhouse-server` to `cmake-build-debug/programs/clickhouse-server`.

#### **Settings -> Build, Execute, Deployment -> CMake** *(ctrl-alt-s)*

* **Build type:** Debug
* **Toolchain:** CMake (3.21.1), Make (gmake), C Compiler (cc), C++ Compiler (c++), Debugger (gdb)
* **Generator:** Ninja
* **CMake options:**
```bash
-DCMAKE_C_COMPILER=/usr/lib/ccache//clang-13
-DCMAKE_CXX_COMPILER=/usr/lib/ccache//clang++-13
-DCMAKE_BUILD_TYPE=Debug
-DENABLE_CLICKHOUSE_ALL=OFF
-DENABLE_CLICKHOUSE_SERVER=ON
-DENABLE_CLICKHOUSE_CLIENT=ON
-DENABLE_LIBRARIES=ON
-DUSE_UNWIND=ON
-DENABLE_UTILS=OFF
-DENABLE_TESTS=OFF
-DWERROR=OFF
-DENABLE_CCACHE=1
```
* **Build directory:** cmake-build-debug (default)

Without explicitly setting `WERROR=OFF` and `ENABLE_LIBRARIES=ON` I run into a number of warnings that cause hard failures, and an inability of the linker to find the `pthreads` library.

**Comment by alexey-milovidov** [comment]
I did not try to use debugger from CLion. You can also use stderr for debugging.
But for compiling and code navigation it works alright.

How to set up CLion:
1. Don't forget to initialize and update submodules.
2. Don't specify any CMake flags. Leave everything by default.

3. Change build type to `RelWithDebInfo` as it is the default build type (and debug build is very rarely needed):
![Screenshot_20220219_172401](https://user-images.githubusercontent.com/18581488/154813549-009a5ccc-f225-42ad-9490-641cc32ee1b9.png)

4. Change `cmake` binary to the binary that you have in your system instead of what is embedded into CLion. It will allow to easily switch between building ClickHouse from command and building inside CLion:
![Screenshot_20220219_172743](https://user-images.githubusercontent.com/18581488/154813551-5c6c891b-12be-4b57-8038-02f034e53f73.png)

5. If `ninja` does not work, replace CLion's ninja it to the ninja that you have in your system.
```
mv /home/milovidov/clion-2021.3.3/bin/ninja/linux/ninja /home/milovidov/clion-2021.3.3/bin/ninja/linux/ninja_
ln -s /usr/bin/ninja /home/milovidov/clion-2021.3.3/bin/ninja/linux/ninja
```

**Comment by alexey-milovidov** [comment]
> The Clickhouse client also warns me that the server is running in debug mode, does this confirm it was built with debug symbols set or is this just more verbose logging?

Yes, this message warns you if you are going to do benchmarks or install this debug version to production by mistake.

**Comment by alexey-milovidov** [comment]
If you suspect that your code has a mistake, you can create a pull request in our repository in "draft" state.
We have automated tests with all sanitizers, automated debug and release builds, cross builds for other platforms, static analysis, fuzzing... and most likely CI will automatically find the error and you only have to check the logs.

**Comment by surahman** [comment]
Thank you, this fixed my build issues. I suspect CLion was using GCC and G++ in spite of the CMake flags.

![Run/Debug Configurations](https://user-images.githubusercontent.com/3979925/154861811-aac62369-6697-4d0e-a3a7-8f3bc9362bfe.png)

I am currently using stderr logging in conjunction with `gdb` and `lldb` on the console. A debugger adds more flexibility to view the variables in the execution frame as well as examine structs and objects. I will reach out to Jetbrains support about why their debugger cannot connect to a target when `gdb` and `lldb` can. I suspect it is a configuration issue.

**Comment by a2888409** [comment]
@surahman you can see this issue had fixed this problem:
fix breakpoint not hit in clion [#35179](https://github.com/ClickHouse/ClickHouse/pull/35179)

there are two way to fix:
1. explicit set cmake option ENABLE_BUILD_PATH_MAPPING to OFF when building
or
2.use ENABLE_BUILD_PATH_MAPPING default ON value, and than:
vim ~/.lldbinit, append following content:

settings set target.load-cwd-lldbinit true
settings append target.source-map . /Users/xxx/Documents/workSpace/ClickHouse

replace ' /Users/xxx/Documents/workSpace/ClickHouse' with your local correct clickhouse root work dir

and reset your breakpoint in CLio , now it get works.

![image](https://user-images.githubusercontent.com/11365761/158136552-867c58bd-6b37-406e-92e7-36ca558305af.png)

**Comment by surahman** [comment]
@a2888409 thank you, this makes life easier than having to jump between CLion and Emacs GDB `many-windows`. Confirming option 1 works with both GDB and LLDB.

Now if only prettyprint of STL containers would work too...
![no_prettyprint](https://user-images.githubusercontent.com/3979925/158240253-7aef7d8f-8f8e-4960-adf7-01fe9f1174ab.png)
